### PR TITLE
[JUJU-3666] Backport cmr secrets from 3.2 branch

### PR DESCRIPTION
**Pull request by wallyworld**
Backport cmr secrets so it is available in the 2 year supported version of Juju.

This involves copying the 3.2 code for the secret facades, the remote relation workers, and some state code to persist the cross model consumer info:
- add new cross model secrets facades
- update secrets manager facade
- new cross model secrets watchers

The current v1 secrets manager facade is retained for compatibility with older 3.1 agents. A new v2 facade is added to handle the new cmr requirements.

## QA steps

run secrets ci tests (known permission error in removing an app owned secret)

plus

bootstrap a 3.1.2 controller
add vault as a backend
add a model and deploy juju-qa-dummy-source and juju-qa-dummy-sink and relate
set vault as the secret backend
create a secret and grant access
ensure secret can be accessed

upgrade the controller to this pr
check that the 3.1.2 model still works for secrets
access the secret
update the secret and check change hook firing

bootstrap a controller off this PR
add vault as a backend
deploy the above charms in a cross model scenario
check that secret access works, secret update workflow works, hooks fired
also check the single model scenario again

### CODE REVIEW

File: tests/suites/secrets_iaas/vault.sh [hunk 1]
```diff
@@ -7,78 +7,11 @@ run_secrets_vault() {
 	juju add-secret-backend myvault vault endpoint="$VAULT_ADDR" token="$VAULT_TOKEN"
 	juju --show-log add-model "$model_name" --config secret-backend=myvault
 
-	juju --show-log deploy easyrsa
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply tests/suites/secrets_iaas/vault.sh:10]: This whole code block was common to the ci tests using mongo backend and vault backend. We really do want to check the same behaviour regardless of backend used. When making the necessary changes in this PR, I was having to do everything twice, once for juju test and again for vault test.

File: tests/suites/secrets_iaas/juju.sh [hunk 2]
```diff
@@ -13,62 +9,68 @@ run_secrets_juju() {
 	wait_for "etcd" "$(idle_condition "etcd" 1 0)"
 	wait_for "active" "$(workload_status "etcd" 0).current"
 
+	echo "Apps deployed, creating secrets"
 	secret_owned_by_easyrsa_0=$(juju exec --unit easyrsa/0 -- secret-add --owner unit owned-by=easyrsa/0)
-	secret_owned_by_easyrsa_0_id=$(echo $secret_owned_by_easyrsa_0 | cut -d: -f 2)
+	secret_owned_by_easyrsa_0_id=${secret_owned_by_easyrsa_0##*/}
```
Comment (ycliuhw) [inline tests/suites/secrets_iaas/juju.sh:14]: Right, so do we need to worry about the compatibility issue of bringing the new URI format back to 3.1?

**Review by ycliuhw — APPROVED** [review]
LGTM, thanks!

File: tests/suites/secrets_iaas/juju.sh (hunk 2 above)
Comment (wallyworld) [reply tests/suites/secrets_iaas/juju.sh:14]: no, uri is opaque

**Pull request merged**